.github/actions/trun/run.ts

```diff
@@ -0,0 +1,133 @@
+import * as esbuild from "./deps/esbuild.ts"
+import { denoPlugin } from "./deps/esbuild_deno_loader.ts"
+import PQueue from "./deps/pqueue.ts"
+import { Browser } from "./deps/puppeteer.ts"
+import { deferred } from "./deps/std/async.ts"
+import { Buffer, readLines } from "./deps/std/io.ts"
+import * as path from "./deps/std/path.ts"
+import { readerFromStreamReader, writeAll } from "./deps/std/streams.ts"
+
+export interface RunOptions {
+  paths: (readonly [dir: string, fileName: string])[]
+  concurrency: number
+  runner: (dir: string, fileName: string) => Promise<void>
+}
+
+export async function run({ paths, concurrency, runner }: RunOptions) {
+  const runQueue = new PQueue({ concurrency })
+  runQueue.addAll(paths.map(([dir, fileName]) => () => runner(dir, fileName)))
+  await runQueue.onIdle()
+}
+
+export interface RunWithBrowserOptions {
+  createBrowser: () => Promise<Browser>
+  importMapURL?: URL
+  results: [fileName: string, exitCode: number][]
+}
+
+export async function runWithBrowser(
+  { createBrowser, importMapURL, results }: RunWithBrowserOptions,
+) {
+  const browser = await createBrowser()
+  const currentDir = new URL(import.meta.url).pathname.split("/").slice(0, -1).join("/")
+  const consoleJS = new TextDecoder().decode(await Deno.readFile(`${currentDir}/console.js`))
+
+  return (async (dir: string, fileName: string) => {
+    console.log(`running ${fileName}`)
+    const outputQueue = new PQueue({ concurrency: 1, autoStart: false })
+
+    outputQueue.add(() => console.log(`${fileName} output:`))
+
+    const page = await browser.newPage()
+    const result = await esbuild.build({
+      plugins: [
+        denoPlugin({
+          importMapURL,
+        }) as any,
+      ],
+      entryPoints: [path.join(dir, fileName)],
+      bundle: true,
+      write: false,
+      format: "esm",
+    })
+
+    const code = wrapCode(`${consoleJS}\n${result.outputFiles[0]?.text!}`)
+
+    const outputBuffer = new Buffer()
+
+    await page.exposeFunction("__trun_injected_log", (...[msg]: [string, number]) => {
+      outputBuffer.writeSync(new TextEncoder().encode(msg))
+    })
+
+    const exit = deferred<number>()
+    await page.exposeFunction("exit", (args: string) => {
+      exit.resolve(Number.parseInt(args))
+    })
+
+    await page.addScriptTag({ content: code, type: "module" })
+
+    const exitCode = await exit
+
+    if (exitCode != 0) {
```

Comment:
```suggestion
    if (exitCode !== 0) {
```

Reply:
does our linter not automatically catch these kinds of errors? typically eslint would catch these and it would get fixed automatically if you have format on saveon.

Reply:
Hmm, not sure if deno lint has a rule to ban `==`. But auto fixing it wouldn't be good, since it does change the semantics.